#Contents List
##Import Libraries
##Import Data - products.csv
##Create a Path
##Import Data - orders_wrangled.csv
##Create a test dataframe
##Create a mixed type column
##Check for mixed types
##Changing mix type to string
##Finding missing values in df_prods dataframe
##View 16 missing values in product_name
##Creating a new dataframe with no missing values
##Finding duplicates
##Create new dataframe without duplicates
##2. Run the df.describe() function on your df_ords dataframe.
##3. Check for mixed types in df_ords
##3. Check for mixed types in df_ords - I could not get the above code to work, so I found new code online that works.
##3. Check for mixed types in df_ords - Jordan's Code
##3. Check for mixed types in df_ords - Jordan's Code 2
##5. Run a check for missing values in your df_ords dataframe.
##View 206,209 missing values in days_since_prior_order
##Create a crosstab to view the 206,209 missing values
##7. Run a check for duplicate values in your df_ords data.
##Export final cleaned products_checked.csv
##Export final cleaned orders_checked.csv

In [1]:
#Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Imported Data set products.csv
df_prods = pd.read_csv(r'C:\Users\Family Computer\Documents\Data Analytics\Instacart Basket Analysis\02 Data\Original Data\products.csv', index_col = False)

In [3]:
##Path
path = r'C:\Users\Family Computer\Documents\Data Analytics\Instacart Basket Analysis'

In [4]:
#Imported Data set orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_wrangled.csv'))

In [5]:
#Create a dataframe
df_test = pd.DataFrame()

In [6]:
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
#Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\Family Computer\AppData\Local\Temp\ipykernel_14836\522087711.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [9]:
#Changing mix type to string
df_test['mix'] = df_test['mix'].astype('str')

In [10]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [11]:
df_test.dtypes

mix    object
dtype: object

In [12]:
#Finding missing values in df_prods dataframe
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
#View 16 missing values in product_name
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
df_prods.shape

(49693, 5)

In [16]:
#Creating a new dataframe with no missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
df_prods_clean.shape

(49677, 5)

In [18]:
#Finding duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [19]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [20]:
df_prods_clean.shape

(49677, 5)

In [21]:
#Create new dataframe without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [22]:
df_prods_clean_no_dups.shape

(49672, 5)

In [23]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  hour_of_day_ordered  days_since_prior_order  
count        3.421083e+06         3.421083e+06            3.214874e+06  
mean         2.776219e+00         1.345202e+01            1.111484e+01  
std          2.046829e+00         4.226088e+00            9.206737e+00  
min          0.000000e+00         0.000000e+00            0.000000e+00  
25%          1.000000e+00         1.000000e+01            4.000000e+00  
50%          3.000000e+00         1.300000e+01            7.000000e+00  
75%          5.000000e+00         1.600000e+01            1.500000e+01  
max          6.000000e+00         2.300000e+01            3.000000e+01

##2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
A few columns appear to have some numbers that I would investigate further: order_number, hour_of_day_ordered, and days_since_prior_order. order_number the min and max are both 1.0000... and the 25% is 5.0, 50% is 1.10, and the 75% is 2.30 - all number sare greater than the max. Also, the order of the 25%, 50%, and 75% are not in sequential order. For the hour_of_day_ordered, the max is 2.30, but there are 24 hours in a day, so I would expect the max to be 24, not 2.3. Finally, for the days_since_prior_order column the max is 3 but the 50% is 7, which is larger than the max. So, this requires further investigation as well. 

In [24]:
##3. Check for mixed types in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

C:\Users\Family Computer\AppData\Local\Temp\ipykernel_14836\1603099487.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Family Computer\AppData\Local\Temp\ipykernel_14836\1603099487.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Family Computer\AppData\Local\Temp\ipykernel_14836\1603099487.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Family Computer\AppData\Local\Temp\ipykernel_14836\1603099487.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).an

In [27]:
##3. Check for mixed types in df_ords - I could not get the above code to work, so I found new code online that works.
for column in df_ords.columns:
    print(column,':',pd.api.types.infer_dtype(df_ords[column]))

Unnamed: 0 : integer
order_id : integer
user_id : integer
order_number : integer
orders_day_of_week : integer
hour_of_day_ordered : integer
days_since_prior_order : floating


In [25]:
##3. Check for mixed types in df_ords - Jordan's Code
for col in df_ords.columns.tolist():
    weird = (df_ords[col].map(type) != df_ords[col].iloc[0].apply(type))
    if weird.any():
        print(col)

AttributeError: 'numpy.int64' object has no attribute 'apply'

In [26]:
##3. Check for mixed types in df_ords - Jordan's Code 2
for col in df_ords.columns:
    mixed_types = df_ords[col].apply(type).nunique() > 1
    if mixed_types:
        print(f"Column '{col}' has mixed data types.")

###3. There is no mixed-type data  - all is either integer or floating.

###4. There is no mixed-type data.

In [28]:
##5. Run a check for missing values in your df_ords dataframe.
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
hour_of_day_ordered            0
days_since_prior_order    206209
dtype: int64

###There are 206,209 missing values in the days_since_prior_order column. A possible explanation is either data corruption or they were never recorded in the first place. 

In [29]:
#View 206,209 missing values in days_since_prior_order
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [30]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         hour_of_day_ordered  days_since_prior_order  
0                          8                     NaN  
11                        11                     NaN  
26                        14                     NaN  
39                        11                     NaN  
45                        12                     NaN  
...                      ...                     ...  
3420930                   12                     NaN  
3420934                   18                     NaN  
3421002                   19                     NaN  
3421019                   15                     NaN  
3421069                   11                     NaN  

[206209 rows x 7 columns]

In [36]:
crosstab = pd.crosstab(df_ords['days_since_prior_order'], df_ords['order_number'], dropna = False)

In [37]:
crosstab.to_clipboard()

##6. Address the missing values using an appropriate method.
It appears that the missing values are not actually missing values, it appears as though it is the first order for each of the 206,209 user_id's. Therefore, I will not do anything with these "missing" values. 

In [ ]:
##7. Run a check for duplicate values in your df_ords data.
df_ords_dups = df_ords[df_ords.duplicated()]

In [ ]:
df_ords_dups

###7. There are no dulpicate rows in the dataframe df_ords.

###8. Address the duplicates using an appropriate method.
There are no duplicate records, therefore nothing needs to be done. 

In [ ]:
##Export final cleaned df_prods
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))


In [ ]:
##Export final cleaned df_ords
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))